rhp/v4/rpc_test.go

```diff
@@ -0,0 +1,1561 @@
+
+	settings, err := rhp4.RPCSettings(context.Background(), transport)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	fundAndSign := &fundAndSign{w, renterKey}
+	renterAllowance, hostCollateral := types.Siacoins(100), types.Siacoins(200)
+	formResult, err := rhp4.RPCFormContract(context.Background(), transport, cm, fundAndSign, cm.TipState(), settings.Prices, hostKey.PublicKey(), settings.WalletAddress, proto4.RPCFormContractParams{
+		RenterPublicKey: renterKey.PublicKey(),
+		RenterAddress:   w.Address(),
+		Allowance:       renterAllowance,
+		Collateral:      hostCollateral,
+		ProofHeight:     cm.Tip().Height + 50,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	revision := formResult.Contract
+
+	cs := cm.TipState()
+	account := proto4.Account(renterKey.PublicKey())
+
+	accountFundAmount := types.Siacoins(25)
+	fundResult, err := rhp4.RPCFundAccounts(context.Background(), transport, cs, renterKey, revision, []proto4.AccountDeposit{
+		{Account: account, Amount: accountFundAmount},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	revision.Revision = fundResult.Revision
+
+	token := proto4.AccountToken{
+		Account:    account,
+		ValidUntil: time.Now().Add(time.Hour),
+	}
+	tokenSigHash := token.SigHash()
+	token.Signature = renterKey.SignHash(tokenSigHash)
+
+	roots := make([]types.Hash256, 10)
+	for i := range roots {
+		// store random sectors on the host
+		data := frand.Bytes(1024)
+
+		// store the sector
+		writeResult, err := rhp4.RPCWriteSector(context.Background(), transport, settings.Prices, token, NewReaderLen(data), 5)
+		if err != nil {
+			t.Fatal(err)
+		}
+		roots[i] = writeResult.Root
+	}
+
+	assertRevision := func(t *testing.T, revision types.V2FileContract, roots []types.Hash256) {
+		t.Helper()
+
+		expectedRoot := proto4.MetaRoot(roots)
+		n := len(roots)
+
+		if revision.Filesize/proto4.SectorSize != uint64(n) {
+			t.Fatalf("expected %v sectors, got %v", n, revision.Filesize/proto4.SectorSize)
+		} else if revision.FileMerkleRoot != expectedRoot {
+			t.Fatalf("expected %v, got %v", expectedRoot, revision.FileMerkleRoot)
+		}
+	}
+
+	// append all the sector roots to the contract
+	appendResult, err := rhp4.RPCAppendSectors(context.Background(), transport, cs, settings.Prices, renterKey, revision, roots)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertRevision(t, appendResult.Revision, roots)
+	revision.Revision = appendResult.Revision
+
+	// randomly remove half the sectors
+	indices := make([]uint64, len(roots)/2)
+	for i, n := range frand.Perm(len(roots))[:len(roots)/2] {
+		indices[i] = uint64(n)
+	}
+	newRoots := append([]types.Hash256(nil), roots...)
+	for i, n := range indices {
+		newRoots[n] = newRoots[len(newRoots)-i-1]
+	}
+	newRoots = newRoots[:len(newRoots)-len(indices)]
+
+	removeResult, err := rhp4.RPCFreeSectors(context.Background(), transport, cs, settings.Prices, renterKey, revision, indices)
+	if err != nil {
+		t.Fatal(err)
+	}
+	assertRevision(t, removeResult.Revision, newRoots)
```

Comment:
This is not the final behavior yet right? Since we wanted to replace the sectors with sentinel values and decrease the size but keep the capacity rather than actually remove the roots.